### PR TITLE
Item 9092: Naming pattern improvements 

### PR DESCRIPTION
**Pull request by XingY**
#### Rationale
Utils to support naming pattern validations and previews

#### Related Pull Requests
* https://github.com/LabKey/labkey-api-js/pull/118
* https://github.com/LabKey/labkey-ui-components/pull/689
* https://github.com/LabKey/platform/pull/2865
* https://github.com/LabKey/sampleManagement/pull/778
* https://github.com/LabKey/biologics/pull/1092

#### Changes
* Add validateNameExpressions, getDomainNamePreviews methods to Domain

### CODE REVIEW

File: src/labkey/Domain.ts [hunk 1]
```diff
@@ -408,6 +408,58 @@ export function save(config: SaveDomainOptions): XMLHttpRequest {
...
+        options: arguments[3],
+        kind: arguments[4],
+        includeNamePreview: arguments[5],
+    } : config;
+
+    return request({
+        url: buildURL('property', 'validateNameExpressions.api', options.containerPath),
+        method: 'POST',
+        success: getCallbackWrapper(getOnSuccess(options), options.scope),
+        failure: getCallbackWrapper(getOnFailure(options), options.scope, true),
```
Comment (labkey-susanh) [inline src/labkey/Domain.ts]: Is `scope` meant to be added to options when parsing the `arguments` array?  I guess this isn't done in other methods that use this pattern, but I'm unclear when the `arguments` option is used and how people know what order to provide those arguments.

File: src/labkey/Domain.ts [hunk 2]
```diff
@@ -408,6 +408,58 @@ export function save(config: SaveDomainOptions): XMLHttpRequest {
...
+        jsonData: {
+            domainDesign: options.domainDesign,
+            options: options.options,
+            kind: options.kind
+        }
+    });
+}
+
+export function getDomainNamePreviews(config: GetDomainDetailsOptions): XMLHttpRequest {
```
Comment (labkey-susanh) [inline src/labkey/Domain.ts:435]: Probably should have some docs on this since it's exported.

File: src/labkey/Domain.ts [hunk 3]
```diff
@@ -408,6 +408,58 @@ export function save(config: SaveDomainOptions): XMLHttpRequest {
...
+
+    let options: GetDomainDetailsOptions = arguments.length > 1 ? {
+        containerPath: arguments[4],
+        failure: arguments[1],
+        queryName: arguments[3],
+        schemaName: arguments[2],
+        success: arguments[0]
+    } : config;
```
Comment (labkey-susanh) [inline src/labkey/Domain.ts]: Ok, really don't know how the order of these arguments is determined.  What am I missing?

**Review by labkey-susanh — APPROVED** [review]
Minor comments.  No need for re-review, though.

File: src/labkey/Domain.ts [hunk 4]
```diff
@@ -408,6 +408,58 @@ export function save(config: SaveDomainOptions): XMLHttpRequest {
...
 
+export interface ValidateDomainNameExpressionOptions extends CreateDomainOptions {
+    includeNamePreview: boolean
+}
+/**
+ * Validate name expression(s) for domain design.
+ */
+export function validateNameExpressions(config: ValidateDomainNameExpressionOptions): XMLHttpRequest {
+
+    let options: ValidateDomainNameExpressionOptions = arguments.length > 1 ? {
```
Comment (labkey-nicka) [inline src/labkey/Domain.ts]: Is this configuration necessary? We only do this for cases where we are providing backwards compatibility for methods that formerly took multiple parameters and now take a configuration object. I would recommend removing this for these new methods.

**Review by labkey-nicka — CHANGES_REQUESTED** [review]

File: src/labkey/Domain.ts (hunk 4 above)
Comment (labkey-nicka) [reply src/labkey/Domain.ts]: @XingY Thanks for adjusting. Looks good.

**Review by labkey-nicka — APPROVED** [review]

**Pull request merged**